components/core/src/ir/LogEventDeserializer.tpp

```diff
@@ -0,0 +1,110 @@
+#ifndef IR_LOGEVENTDESERIALIZER_TPP
+#define IR_LOGEVENTDESERIALIZER_TPP
+
+#include <vector>
+
+#include <json/single_include/nlohmann/json.hpp>
+
+#include "../ffi/ir_stream/decoding_methods.hpp"
+#include "LogEventDeserializer.hpp"
+
+namespace ir {
+template <typename encoded_variable_t>
+auto LogEventDeserializer<encoded_variable_t>::create(ReaderInterface& reader)
+        -> BOOST_OUTCOME_V2_NAMESPACE::std_result<LogEventDeserializer<encoded_variable_t>> {
+    ffi::ir_stream::encoded_tag_t metadata_type{0};
+    std::vector<int8_t> metadata;
+    auto ir_error_code = ffi::ir_stream::decode_preamble(reader, metadata_type, metadata);
+    if (ffi::ir_stream::IRErrorCode_Success != ir_error_code) {
+        switch (ir_error_code) {
+            case ffi::ir_stream::IRErrorCode_Incomplete_IR:
+                return std::errc::result_out_of_range;
+            case ffi::ir_stream::IRErrorCode_Corrupted_IR:
+            default:
+                return std::errc::protocol_error;
+        }
+    }
+
+    if (ffi::ir_stream::cProtocol::Metadata::EncodingJson != metadata_type) {
+        return std::errc::protocol_not_supported;
+    }
+
+    // Parse metadata and validate version
+    auto metadata_json = nlohmann::json::parse(metadata, nullptr, false);
+    if (metadata_json.is_discarded()) {
+        return std::errc::protocol_error;
+    }
+    auto version_iter = metadata_json.find(ffi::ir_stream::cProtocol::Metadata::VersionKey);
+    if (metadata_json.end() == version_iter || false == version_iter->is_string()) {
+        return std::errc::protocol_error;
+    }
+    auto metadata_version = version_iter->get_ref<nlohmann::json::string_t&>();
+    if (static_cast<char const*>(ffi::ir_stream::cProtocol::Metadata::VersionValue)
+        != metadata_version)
+    {
+        return std::errc::protocol_not_supported;
+    }
+
+    if constexpr (std::is_same_v<encoded_variable_t, ffi::eight_byte_encoded_variable_t>) {
+        return LogEventDeserializer<encoded_variable_t>{reader};
+    } else if constexpr (std::is_same_v<encoded_variable_t, ffi::four_byte_encoded_variable_t>) {
```

Comment:
technically you don't need the else here and on line 64?

Reply:
We did since constexpr is not evaluated in control flow; this is no longer an issue since I moved the implementation into the CPP and used explicit specializations.